scipy/optimize/_isotonic.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from ._pava_pybind import pava
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+
+__all__ = ["isotonic_regression"]
+
+
+def isotonic_regression(
+    y: npt.ArrayLike,
+    weights: npt.ArrayLike | None = None,
+    increasing: bool = True,
+):
+    r"""Nonparametric isotonic regression.
+
+    A monotonically increasing array `x` with the same length as `y` is
+    calculated by the pool adjacent violators algorithm (PAVA), see [1]_.
+    See the Notes section for more details.
+
+    Parameters
+    ----------
+    y : (N,) array_like
+        Response variable.
+    weights : (N,) array_like or None
+        Case weights.
+    increasing : bool
+        If True, fit monotonic increasing, i.e. isotonic, regression.
+        If False, fit a monotonic decreasing, i.e. antitonic, regression.
+
+    Returns
+    -------
+    x : (N,) array_like
+        Isotonic regression solution, i.e. an increasing (or decresing) array
+        of the same length than y.
+    wx : (B,) array_like
+        Sum of case weights for all blocks / pools B.
+    r : (B+1,) array_like
+        Array of indices with the start position of each block / pool B.
+        For the j-th block, all values of ``x[r[j]:r[j+1]]`` are the same.
+
+    Notes
+    -----
+    Given data :math:`y` and case weights :math:`w`, the isotonic regression
+    solves the following optimization problem:
+
+    .. math::
+
+        \operatorname{argmin}_{x_i} \sum_i w_i (y_i - x_i)^2 \quad
+        \text{subject to } x_i \leq x_j \text{ whenever } i \leq j \,.
+
+    For every input value :math:`y_i`, it generates an interpolated value
+    :math:`x_i` which are increasing. This is accomplished by the PAVA.
+    The solution consists of pools or blocks, i.e. neighboring elements of
+    :math:`x`, e.g. :math:`x_i` and :math:`x_{i+1}`, that all have the same
+    value.
+
+    Most interestingly, the solution stays the same if the squared loss is
+    replaced by the wide class of Bregman functions which are the unique
+    class of strictly consistent scoring functions for the mean, see [2]_
+    and references therein.
```

Comment:
That's for experts because even experts might not know this and it is very valuable information. It also shows that the solution is a very general and special one.